wallet-new/src/Cardano/Wallet/API/V1/Handlers/Internal.hs

```diff
@@ -0,0 +1,51 @@
+module Cardano.Wallet.API.V1.Handlers.Internal
+    ( computeUtxoStatistics
+    )  where
+
+
+import           Prelude
+
+import           Cardano.Wallet.API.V1.Types
+
+import qualified Control.Foldl as L
+import qualified Data.Map.Strict as MS
+import qualified Data.Text as T
+import           Data.Word
+
+
+computeUtxoStatistics :: [Word64] ->  UtxoStatistics
+computeUtxoStatistics xs = L.fold (summarizeUtxoStatistics $ generateBounds Log10) xs
+
+-- | Using foldl library enable as to capture a number of aggregations in one pass. This thanks to L.Fold being an Applicative
+summarizeUtxoStatistics :: [Word64] -> L.Fold Word64 UtxoStatistics
+summarizeUtxoStatistics bounds = UtxoStatistics
+  <$> populateBuckets bounds
+  <*> L.sum
+
+-- | Buckets boundaries can be constructed in different way
+data BoundType = Log10 | Haphazard
+
+generateBounds :: BoundType -> [Word64]
+generateBounds bType =
+    case bType of
+        Log10 -> ( map (\toPower -> 10^toPower :: Word64) [(1::Word64)..16] ) ++ [45 * (10^(15::Word64))]
+        Haphazard -> [10, 100, 1000, 10000]
+
+
+populateBuckets ::  [Word64] ->  L.Fold Word64 [HistogramBar]
+populateBuckets bounds =
+    case bounds of
+        (x:_) -> L.Fold (addCountInBuckets x) (initalizeMap bounds) (fmap (\pair -> HistogramBarCount (T.pack $ show $ fst pair, snd pair) ) . MS.toList)
+        _ -> error "populateBuckets has to be powered with nonempty bounds"
```

Comment:
Then, use a `NonEmpty Word64` instead of a `[Word 64]` :D !

https://hackage.haskell.org/package/base-4.11.1.0/docs/Data-List-NonEmpty.html